lib/cn/kblock_builder.c

```diff
@@ -967,10 +968,14 @@ kblock_finish(struct kblock_builder *bld, struct wbb *ptree)
         goto errout;
     }
 
+    /* Set preallocate flag if this kblock's write length >= 90% of the mblock size */
+    if (wlen >= ((kblocksz * 9) / 10))
+        flags |= MPOOL_MBLOCK_PREALLOC;
+
```

Comment:
Is the 90% threshold based on experiments?

Reply:
Nope, I just picked a number for now so that the space-amp doesn't look worse.

Reply:
you could use (kblocksize * 921 / 1024) to run much faster!

Reply:
although the compiler should turn this into a multiply+shift without your help, since you're using constants...

Reply:
Got curious about this: https://godbolt.org/z/vvKbq4xs6

Reply:
that's pretty sneaky using the address calculation hardware, lemme turn that into a compare against a power of two and see what it generates...

Reply:
So, it'll try to use the address decoder to do the math when given small numbers, but that restricts it to multples of 2/4/8 with an addition ot the addend or the result to get different results.  So using 90/100 or 921/1024 results in a different approach which looks more expensive.  But the literal 1000000 and return of x*9 completely change the problem.

https://godbolt.org/z/W493a8W9P

Reply:
I think my opinion on the matter is premature optimization, and if this one math expression is our bottleneck then we are really good at our jobs :).

Reply:
Death by a thousand cuts...

Reply:
I think a comment is at least necessary. 9/10 is easy to recognize. 921/1024 is not.

Reply:
Well, it turns out that 9/10 looks more optimial than 921/1024.  However, that's due to an optimization that can leverage the address computation hardware.  If it were 93/100 then it could not leverage that trick.  So may as well leave it as is for now..

Reply:
Seems like you guys had some interesting conversation here!